website/blog/2023-07-14-Local-LLMs/index.mdx

```diff
@@ -0,0 +1,107 @@
+---
+title: Use flaml.autogen for local LLMs
+authors: jialeliu
+tags: [LLM, FLAMLv2]
+---
+**TL;DR:**
+We demonstrate how to use flaml.autogen for local LLM application. As an example, we will initiate an endpoint using [FastChat](https://github.com/lm-sys/FastChat) and perform inference on [ChatGLMv2-6b](https://github.com/THUDM/ChatGLM2-6B).
+
+## Preparations
+
+### Clone FastChat
+
+FastChat provides OpenAI-compatible APIs for its supported models, so you can use FastChat as a local drop-in replacement for OpenAI APIs. However, it's code needs minor modification in order to function properly.
+
+```bash
+git clone https://github.com/lm-sys/FastChat.git
+cd FastChat
+```
+
+According to this [issue](https://github.com/lm-sys/FastChat/issues/1641), comment out all the line containing `finish_reason` in `fastchat/protocol/api_protocal.py` and `fastchat/protocol/openai_api_protocol.py` will fix the errors. This is what modified code looks like:
+
+```python
+class CompletionResponseChoice(BaseModel):
+    index: int
+    text: str
+    logprobs: Optional[int] = None
+    # finish_reason: Optional[Literal["stop", "length"]]
+
+class CompletionResponseStreamChoice(BaseModel):
+    index: int
+    text: str
+    logprobs: Optional[float] = None
+    # finish_reason: Optional[Literal["stop", "length"]] = None
+```
+
+### Download checkpoint
+
+ChatGLM-6B is an open bilingual language model based on General Language Model (GLM) framework, with 6.2 billion parameters. ChatGLM2-6B is its second-generation version.
+
+Before downloading from HuggingFace Hub, you need to have Git LFS [installed](https://docs.github.com/en/repositories/working-with-files/managing-large-files/installing-git-large-file-storage).
+
+```bash
+git clone https://huggingface.co/THUDM/chatglm2-6b
+```
+
+## Initiate server
+
+First, launch the controller
+
+```bash
+python -m fastchat.serve.controller
+```
+
+Then, launch the model worker(s)
+
+```bash
+python -m fastchat.serve.model_worker --model-path chatglm2-6b
+```
+
+Finally, launch the RESTful API server
+
+```bash
+python -m fastchat.serve.openai_api_server --host localhost --port 8000
+```
+
+## Interact with model using `oai.Completion`
+
+Now the models can be directly accessed through openai-python library as well as `flaml.oai.Completion`.
+
+```python
+from flaml import oai
+
+# create a text completion request
+response = oai.Completion.create(
+    config_list=[
+        {
+            "model": "chatglm2-6b",
+            "api_base": "http://localhost:8000/v1",
```

Comment:
Can the api be hosted remotely on a server?

Reply:
I think so. Change localhost to the IP address of the server will make it work.

Reply:
Could you document that too?

Reply:
Unfortunately this method is only doable on local machines. Hosting api remotely would require some other tools other than fastchat.